### PR TITLE
Dataset "schema" v0.3

### PR DESCRIPTION
**Pull request by ravwojdyla**
Re: #43 
Resurrect: #124 (please read it for context)

~Depends on #274~

### CODE REVIEW

File: sgkit/stats/aggregation.py [hunk 1]
```diff
@@ -216,14 +241,30 @@ def allele_frequency(ds: Dataset) -> Dataset:
...
     """Compute quality control variant statistics from genotype calls.
 
     Parameters
     ----------
     ds
         Genotype call dataset such as from
         `sgkit.create_genotype_call_dataset`.
+    call_genotype
+        Input variable name holding call_genotype.
+        As defined by `sgkit.variables.call_genotype`.
```
Comment (ravwojdyla) [inline sgkit/stats/aggregation.py:261]: We should probably decide on how we want to do this linking, and what kind of information goes into method doc vs variable docs. I would actually suggest we do this in a separate PR, to keep review slim. But if people feel strong to do it here, fine as well.

File: sgkit/variables.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+"""TODO"""
+call_genotype_probability = ArrayLikeSpec("call_genotype_probability", kind="f", ndim=3)
+"""TODO"""
+call_genotype_probability_mask = ArrayLikeSpec(
+    "call_genotype_probability_mask", kind="b", ndim=3
+)
+"""TODO"""
+genotype_counts = ArrayLikeSpec("genotype_counts", ndim=2, kind="i")
+"""
+Genotype counts. Must correspond to an (`N`, 3) array where `N` is equal
```
Comment (ravwojdyla) [inline sgkit/variables.py:57]: FYI, the 1st sentence appears on the front page of the variable, the rest appears on the variable specific page.

File: sgkit/variables.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+@dataclass(frozen=True)
+class ArrayLikeSpec(Spec):
+    """ArrayLike type spec"""
+
+    kind: Union[None, str, Set[str]] = None
+    ndim: Union[None, int, Set[int]] = None
+
+
+call_genotype = ArrayLikeSpec("call_genotype", kind="i", ndim=3)
```
Comment (ravwojdyla) [inline sgkit/variables.py:22]: We should decide if we want to list this in a single namespace, group these, and or what order do we want them documented. Also would be good to have bidirectional links between variables and methods.

File: sgkit/variables.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+call_dosage = ArrayLikeSpec("call_dosage", kind="f", ndim=2)
+"""Dosages, encoded as floats, with NaN indicating a missing value."""
+call_dosage_mask = ArrayLikeSpec("call_dosage_mask", kind="b", ndim=2)
+"""TODO"""
+call_genotype_probability = ArrayLikeSpec("call_genotype_probability", kind="f", ndim=3)
+"""TODO"""
+call_genotype_probability_mask = ArrayLikeSpec(
+    "call_genotype_probability_mask", kind="b", ndim=3
+)
+"""TODO"""
```
Comment (ravwojdyla) [inline sgkit/variables.py:54]: There are some TODO's still where the doc was missing at the origin.

File: sgkit/stats/aggregation.py (hunk 1 above)
Comment (ravwojdyla) [reply sgkit/stats/aggregation.py]: FYI added links between methods and variables in the documentation.

File: sgkit/model.py [hunk 5]
```diff
@@ -83,7 +79,9 @@ def create_genotype_call_dataset(
         check_array_like(variant_id, kind={"U", "O"}, ndim=1)
         data_vars["variant_id"] = ([DIM_VARIANT], variant_id)
     attrs: Dict[Hashable, Any] = {"contigs": variant_contig_names}
-    return xr.Dataset(data_vars=data_vars, attrs=attrs)
+    return variables.validate(
+        xr.Dataset(data_vars=data_vars, attrs=attrs), *data_vars.keys()
```
Comment (eric-czech) [inline sgkit/model.py:83]: Any reason not to have a `validate` overload that checks against all sgkit.variables by default (to avoid `*data_vars.keys`)?

File: sgkit/stats/hwe.py [hunk 6]
```diff
@@ -140,6 +144,12 @@ def hardy_weinberg_test(
         where `N` is equal to the number of variants and the 3 columns contain
         heterozygous, homozygous reference, and homozygous alternate counts
         (in that order) across all samples for a variant.
+    call_genotype
+        Input variable name holding call_genotype.
+        As defined by :data:`sgkit.variables.call_genotype`.
```
Comment (eric-czech) [inline sgkit/stats/hwe.py:149]: nit: This isn't a complete sentence which will be a little odd given how common "As defined by X." is.  I would suggest "Defined by X." instead.

File: sgkit/variables.py (hunk 3 above)
Comment (eric-czech) [reply sgkit/variables.py:22]: +1 to making it possible to see what methods use or produce a variable.

**Review by eric-czech — CHANGES_REQUESTED** [review]
LGTM other than a couple minor things.  Thanks for cleaning up all those signatures!  Much more consistent now.

File: sgkit/stats/aggregation.py [hunk 7]
```diff
@@ -51,16 +52,20 @@ def count_alleles(g: ArrayLike, _: ArrayLike, out: ArrayLike) -> None:
             out[a] += 1
 
 
-def count_call_alleles(ds: Dataset, merge: bool = True) -> Dataset:
+def count_call_alleles(
+    ds: Dataset, *, call_genotype: str = "call_genotype", merge: bool = True
```
Comment (tomwhite) [inline sgkit/stats/aggregation.py:56]: Nice change here to make variable name arguments (and `merge`) keyword only.

**Review by tomwhite — CHANGES_REQUESTED** [review]
This looks great - thanks @ravwojdyla. I have a few suggestions (mostly minor), but this should be ready to merge soon.

One thing worth mentioning is that users writing their own methods do not have to use the validation if they don't want to (in fact, it's not really a part of the public API so it's not easy to use). I think that's fine - we want to encourage users to take advantage of xarray APIs as needed - and this work is to make sgkit's methods more internally consistent and well-documented.

File: sgkit/stats/aggregation.py [hunk 8]
```diff
@@ -113,17 +119,24 @@ def count_call_alleles(ds: Dataset, merge: bool = True) -> Dataset:
             )
         }
     )
-    return conditional_merge_datasets(ds, new_ds, merge)
+    return variables.validate(
+        conditional_merge_datasets(ds, new_ds, merge), "call_allele_count"
```
[the remark this answers is not in the document]
Comment (ravwojdyla) [reply sgkit/stats/aggregation.py]: @tomwhite this is outdated via 47eef29

File: sgkit/variables.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+corresponding to the number of non-missing occurrences of each allele.
+"""
+variant_allele_count = ArrayLikeSpec("variant_allele_count", ndim=2, kind="u")
+"""
+Variant allele counts. With shape (variants, alleles) and values
+corresponding to the number of non-missing occurrences of each allele.
+"""
+variant_hwe_p_value = ArrayLikeSpec("variant_hwe_p_value", kind="f")
+"""P values from HWE test for each variant as float in [0, 1]."""
+variant_beta = ArrayLikeSpec("variant_beta")
```
[the remark this answers is not in the document]
Comment (ravwojdyla) [reply sgkit/variables.py:77]: For this and the other "naming" comments below + all the TODOs in the docs, I suggest we open a separate issue and follow up in separate PR(s). One of the nice side effects of having all those variables in the same place is that we can see the inconsistencies etc. wdyt @tomwhite ?

[the remark this answers is not in the document]
Comment (tomwhite) [reply sgkit/variables.py]: Yes, fine by me

File: sgkit/stats/aggregation.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tomwhite) [reply sgkit/stats/aggregation.py]: OK, but the question still remains where the dataset is created. A typo there won't cause validation to fail will it?

[the remark this answers is not in the document]
Comment (ravwojdyla) [reply sgkit/stats/aggregation.py]: @tomwhite a typo there would definitely cause a validation error (which is intentional/good). Whether we should reference variable or use strings is a good question, but more from a "dev practices" standpoint. I guess I lean towards referencing it, BUT it would mean that you need to for example say: `variables.pc_relate_phi.default_name` instead of `"pc_relate_phi"`. We could shorten it to `variables.pc_relate_phi`, if we provide some magic methods (hash/str). wdyt?

[the remark this answers is not in the document]
Comment (tomwhite) [reply sgkit/stats/aggregation.py]: Yes, I would probably lean towards referencing the variable too.

File: sgkit/variables.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ravwojdyla) [reply sgkit/variables.py]: @tomwhite so I just remembered something, this would be going against some early comments in https://github.com/pystatgen/sgkit/issues/17, what do you think? also ping @eric-czech

[the remark this answers is not in the document]
Comment (tomwhite) [reply sgkit/variables.py]: @ravwojdyla do you mean the comment about `variables.pc_relate_phi.default_name` vs `"pc_relate_phi"`? OK, happy to leave as a string.

[the remark this answers is not in the document]
Comment (ravwojdyla) [reply sgkit/variables.py]: @tomwhite right, wrong comment section, sorry, this should have been in https://github.com/pystatgen/sgkit/pull/276#discussion_r495782207

Let's see what @eric-czech thinks. Other than that, afaiu @tomwhite this is ready for another review.

File: sgkit/stats/aggregation.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ravwojdyla) [reply sgkit/stats/aggregation.py]: for posterity, I've commented on the wrong comment section, copying over here:

> @tomwhite so I just remembered something, this would be going against some early comments in #17, what do you think? also ping @eric-czech

File: sgkit/variables.py (hunk 3 above)
Comment (jeromekelleher) [reply sgkit/variables.py:55]: I agree with @tomwhite here

**Review by jeromekelleher — APPROVED** [review]
LGTM!

File: sgkit/stats/aggregation.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (eric-czech) [reply sgkit/stats/aggregation.py]: I think using sgkit.variables for creating the output datasets sounds good.  The closest thing to an edge case I can think of right now is a function that generates a mask or mean imputes one of several possible input variables.  Perhaps validation could eventually be more dynamic for those cases.  Otherwise, I can't imagine where using a constant would be awkward.

[the remark this answers is not in the document]
Comment (ravwojdyla) [reply sgkit/stats/aggregation.py]: @tomwhite @eric-czech ok, this PR is now using `sgkit.variables` pretty much everywhere we need a variable name. I did not like the overly verbose `sgkit.variables.<varname>.default_name`, and making `Spec` hashable and acting like a string introduces complexity and will likely lead to issues in xr (see [here](https://github.com/ravwojdyla/sgkit/commit/b6197a94d155dba0099aa09a7030d9f2006bac20)), so I've opted the solution you can see in 6271e89.

**Comment by ravwojdyla** [comment]
FYI rebased to remove the conflict.

**Review by tomwhite — APPROVED** [review]

**Review by eric-czech — APPROVED** [review]
Looks good!

**Comment by jeromekelleher** [comment]
I've marked this auto-merge @ravwojdyla, should merge once you've cleared up the conflicts.
